src/runtime/threading_backend.cc

```diff
@@ -34,13 +34,63 @@
 #endif
 #if defined(__hexagon__)
 #include <dlfcn.h>
+#include <qurt.h>
+#include <stdlib.h>
+#define HEXAGON_STACK_SIZE 65536
+#define HEXAGON_STACK_ALIGNMENT 32
 #endif
 #include <algorithm>
 #include <thread>
 #define CURRENT_THREAD_HANDLE (static_cast<std::thread::native_handle_type>(0))
 namespace tvm {
 namespace runtime {
 namespace threading {
+#ifdef __hexagon__
```

Comment:
wondering if it's possible to do this with templating instead of with `#ifdef __hexagon__`?

Reply:
To a certain extent, although it's not clear it's worth the effort. As far as I can tell are a couple options:

- Have multiple classes that implement the ThreadGroup::Impl interface -- one for pthreads, and one for QuRT threads. @csullivan was concerned that this might lead to duplicated code, making it more fragile to maintain.
- Parameterizing ThreadPool::Impl on the underlying thread type. However, we'd still need #ifdefs to avoid calling SetAffinity, which isn't available on Hexagon.

Yield() also doesn't work as-is on Hexagon, so we'd need a fix for that as well.

Reply:
One approach would be to introduce 

```
template <typename ThreadType>
class ThreadGroup::Impl;
```

and then wrap both std::thread and QuRT thread into types with a common interface that `ThreadGroup::Impl` calls into. 

The thing I got hung up on there is the runtime dispatch. Right now it should be doable from the device type -- when it is `kDLHexagon` dispatch to `ThreadGroup::Impl<QuRTThreadInterface>` when it's `kDLCPU` dispatch to `TheadGroup::Impl<StdThreadInterface>`. However there is impetus to move Hexagon fully over to kDLCPU -- wherein we could no longer do runtime dispatch based on the device type.

Reply:
Then the only other options for dispatch I see are then 

(1) TVM compile-time dispatch -- during codegen, do something specific based on the target 
(2) Build time dispatch via either preprocessor or through changes to the build system to conditionally include one translation unit over another.

Reply:
I think that abstracting thread control would be the best approach here.  The runtime built for a particular target could then contain the implementation provided by that target (e.g. via template specialization), and there wouldn't be any run-time dependency on device kind.

Reply:
I've refactored this PR to split up pthread and qurt threading implementations. This introduces the following changes:

- ThreadGroup::Impl is now an abstract class.
- A ThreadGroupImplTemplate class is introduced with common code between pthread and qurt implementations. It is a subclass of ThreadGroup::Impl. (AFAICT, you can't have a pointer to an unspecialized template class. We need a pointer to a concrete type for ThreadGroup to call.)
- ThreadGroupPosixImpl now contains the bulk of the code from ThreadGroup::Impl. It inherits from ThreadGroupImplTemplate<std::thread>. This is in a new file src/runtime/posix/threading_posix.cc.
- ThreadGroupHexagonImpl inherits from ThreadGroupImplTemplate<QuRTThread>, which are both defined in a new file, src/runtime/hexagon/threading_hexagon.cc.
- There are now two different versions of Yield()
- CMakeLists.txt has been modified to either include src/runtime/posix/threading_posix.cc or src/runtime/hexagon/threading_hexagon.cc, depending on whether we're building for Hexagon.

Honestly, this seems like a pretty ugly solution, especially when threading_backend.cc is already littered with #ifdefs for various platforms. Thoughts?

Reply:
@areusch @kparzysz-quic PTAL. I'm inclined to take @supersat's recommendation if this approach is too ugly. But we'd like to keep forward progress on this one. Thanks!